pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -165,46 +190,305 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		return nil
+	}
+
+	allNodeNames := make([]string, 0, len(nodeInfoMap))
+	for name := range nodeInfoMap {
+		allNodeNames = append(allNodeNames, name)
+	}
+
+	var lock sync.Mutex
+	var firstError error
+
+	topologyPairsPodSpreadMap := newTopologyPairsPodSpreadMap()
+
+	appendTopologyPairsMaps := func(toAppend *topologyPairsMaps) {
+		lock.Lock()
+		topologyPairsPodSpreadMap.appendMaps(toAppend)
+		lock.Unlock()
+	}
+	catchError := func(err error) {
+		lock.Lock()
+		if firstError == nil {
+			firstError = err
+		}
+		lock.Unlock()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	processNode := func(i int) {
+		nodeInfo := nodeInfoMap[allNodeNames[i]]
+		node := nodeInfo.Node()
+		if node == nil {
+			catchError(fmt.Errorf("node not found"))
+			return
+		}
+		// check if current node satisfies nodeSelector / hard nodeAffinity
+		if !podMatchesNodeSelectorAndAffinityTerms(pod, node) {
+			return
+		}
+		// ensure current node's labels contains all topologyKeys in 'constraints'
+		if !nodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
+			return
+		}
+
+		nodeTopologyMaps := newTopologyPairsMaps()
+		// nodeInfo.Pods() can be empty; or all pods don't fit
+		for _, existingPod := range nodeInfo.Pods() {
+			ok, err := podMatchesAllSpreadConstraints(existingPod, pod.Namespace, constraints)
+			if err != nil {
+				catchError(err)
+				cancel()
+				return
+			}
+			if ok {
+				for _, constraint := range constraints {
+					// constraint.TopologyKey is already guaranteed to be present
+					pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
+					nodeTopologyMaps.addTopologyPair(pair, existingPod)
+				}
+			}
+		}
+		// If needed, append "headless" topology pair
+		// For example, on node-x, there is no pod matching spread constraints
+		// but node-x should be also considered as a match (with match number 0)
+		// i.e. <node: node-x>: {}
+		for _, constraint := range constraints {
+			// constraint.TopologyKey is already guaranteed to be present
+			pair := topologyPair{
+				key:   constraint.TopologyKey,
+				value: node.Labels[constraint.TopologyKey],
+			}
+			// addHeadlessTopologyPair is a non-op if other pods match this pair
+			nodeTopologyMaps.addHeadlessTopologyPair(pair)
+		}
+
+		appendTopologyPairsMaps(nodeTopologyMaps)
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processNode)
+
+	// calculate min match for each topology pair
+	topologyPairsPodSpreadMap.minMatchMap = make(map[string]int32, len(constraints))
+	for _, constraint := range constraints {
+		topologyPairsPodSpreadMap.minMatchMap[constraint.TopologyKey] = MaxInt32
+	}
+	for pair, podSet := range topologyPairsPodSpreadMap.topologyPairToPods {
+		// TODO(Huang-Wei): short circuit to skip non-visited <topologyKey: any value> pairs
+		// if we see 0 as min match of the topologyKey
+		if l := int32(len(podSet)); l < topologyPairsPodSpreadMap.minMatchMap[pair.key] {
```

Comment:
Should we check that len(podSet) > 0 ?

Reply:
Nope. If l == 0, it means there is a topology domain having 0 pod matching incoming pod's topologySpreadConstrains.